modulus/metrics/diffusion/loss.py

```diff
@@ -0,0 +1,376 @@
+# limitations under the License.
+
+
+"""Loss functions used in the paper
+"Elucidating the Design Space of Diffusion-Based Generative Models"."""
+
+from typing import Union
+
+import torch
+
+
+class VPLoss:
+    """
+    Loss function corresponding to the variance preserving (VP) formulation.
+
+    Parameters
+    ----------
+    beta_d: float, optional
+        Coefficient for the diffusion process, by default 19.9.
+    beta_min: float, optional
+        Minimum bound, by defaults 0.1.
+    epsilon_t: float, optional
+        Small positive value, by default 1e-5.
+
+    Note:
+    -----
+    Reference: Song, Y., Sohl-Dickstein, J., Kingma, D.P., Kumar, A., Ermon, S. and
+    Poole, B., 2020. Score-based generative modeling through stochastic differential
+    equations. arXiv preprint arXiv:2011.13456.
+
+    """
+
+    def __init__(
+        self, beta_d: float = 19.9, beta_min: float = 0.1, epsilon_t: float = 1e-5
+    ):
+        self.beta_d = beta_d
+        self.beta_min = beta_min
+        self.epsilon_t = epsilon_t
+
+    def __call__(self, net, images, labels, augment_pipe=None):
+        rnd_uniform = torch.rand([images.shape[0], 1, 1, 1], device=images.device)
+        sigma = self.sigma(1 + rnd_uniform * (self.epsilon_t - 1))
+        weight = 1 / sigma**2
+        y, augment_labels = (
+            augment_pipe(images) if augment_pipe is not None else (images, None)
+        )
+        n = torch.randn_like(y) * sigma
+        D_yn = net(y + n, sigma, labels, augment_labels=augment_labels)
+        loss = weight * ((D_yn - y) ** 2)
+        return loss
+
+    def sigma(
+        self, t: Union[float, torch.Tensor]
+    ):  # NOTE: also exists in preconditioning
+        """
+        Compute the sigma(t) value for a given t based on the VP formulation.
+
+        The function calculates the noise level schedule for the diffusion process based
+        on the given parameters `beta_d` and `beta_min`.
+
+        Parameters
+        ----------
+        t : Union[float, torch.Tensor]
+            The timestep or set of timesteps for which to compute sigma(t).
+
+        Returns
+        -------
+        torch.Tensor
+            The computed sigma(t) value(s).
+        """
+        t = torch.as_tensor(t)
+        return ((0.5 * self.beta_d * (t**2) + self.beta_min * t).exp() - 1).sqrt()
+
+
+class VELoss:
+    """
+    Loss function corresponding to the variance exploding (VE) formulation.
+
+    Parameters
+    ----------
+    sigma_min : float
+        Minimum supported noise level, by default 0.02.
+    sigma_max : float
+        Maximum supported noise level, by default 100.0.
+
+    Note:
+    -----
+    Reference: Song, Y., Sohl-Dickstein, J., Kingma, D.P., Kumar, A., Ermon, S. and
+    Poole, B., 2020. Score-based generative modeling through stochastic differential
+    equations. arXiv preprint arXiv:2011.13456.
+    """
+
+    def __init__(self, sigma_min: float = 0.02, sigma_max: float = 100.0):
```

Comment:
Same comments as for `VPLoss`, documentation needs to be improved.

Reply:
Added documentation for the `__call__` method